src/main/java/ch/njol/skript/expressions/ExprName.java

```diff
@@ -97,336 +96,151 @@
-		if (changeType == ITEM) {
-			if (Slot.class.isAssignableFrom(getExpr().getReturnType())) {
-				for (Slot s : (Slot[]) getExpr().getArray(e)) {
-					ItemStack i = s.getItem();
-					type.set(i, name);
-					s.setItem(i);
+		String name = delta != null ? (String) delta[0] : null;
+		for (Object o : getExpr().getArray(e)) {
+			if (o instanceof Player) {
+				switch (mark) {
+					case 2: ((Player) o).setDisplayName(name != null ? name + ChatColor.RESET : ((Player) o).getName());
+					case 3: ((Player) o).setPlayerListName(name); // If name is null, then the player's name will be used.
 				}
-			} else {
-				Object i = getExpr().getSingle(e);
-				if (i instanceof ItemType) {
-					type.set(i, name);
-					getExpr().change(e, new ItemType[] {(ItemType) i}, ChangeMode.SET);
+			} else if (o instanceof Entity) {
+				((Entity) o).setCustomName(name);
+				if (mark == 2 || mode == ChangeMode.RESET) // Using "display name"
+					((Entity) o).setCustomNameVisible(name != null);
+				if (o instanceof LivingEntity)
+					((LivingEntity) o).setRemoveWhenFarAway(name == null);
+			} else if (o instanceof ItemType) {
+				ItemType i = (ItemType) o;
+				ItemMeta m = i.getItemMeta();
+				m.setDisplayName(name);
+				i.setItemMeta(m);
+			} else if (o instanceof Inventory ) {
+				Inventory inv = (Inventory) o;
+
+				if (inv.getViewers().isEmpty())
 					return;
-				}
+				// Create a clone to avoid a ConcurrentModificationException
+				List<HumanEntity> viewers = new ArrayList<>(inv.getViewers());
 
-				if (!(i instanceof ItemStack) && !(i instanceof Slot))
+				InventoryType type = inv.getType();
+				if (!type.isCreatable())
 					return;
-				ItemStack is = i instanceof Slot ? ((Slot) i).getItem() : (ItemStack) i;
-				type.set(is, name);
-				if (i instanceof Slot)
-					((Slot) i).setItem(is);
-				else if (ChangerUtils.acceptsChange(getExpr(), ChangeMode.SET, ItemStack.class))
-					getExpr().change(e, new Object[] {i}, ChangeMode.SET);
-				else
-					getExpr().change(e, new ItemType[] {new ItemType((ItemStack) i)}, ChangeMode.SET);
-			}
-		} else {
-			for (Object o : getExpr().getArray(e)) {
-				if (o instanceof Entity || o instanceof Inventory)
-					type.set(o, name);
+				if (name == null)
+					name = type.getDefaultTitle();
+
+				Inventory copy;
+				if (type == InventoryType.CHEST) {
+					copy = Bukkit.createInventory(inv.getHolder(), inv.getSize(), name);
+				} else {
+					copy = Bukkit.createInventory(inv.getHolder(), type, name);
+				}
+				copy.setContents(inv.getContents());
+				viewers.forEach(viewer -> viewer.openInventory(copy));
+			} else if (o instanceof Slot) {
+				Slot s = (Slot) o;
+				ItemStack is = s.getItem();
+				if (is != null && is.hasItemMeta()) {
+					ItemMeta m = is.getItemMeta();
+					m.setDisplayName(name);
+					is.setItemMeta(m);
+					s.setItem(is);
+				}
 			}
 		}
 	}
 
 	@Override
-	public Class<String> getReturnType() {
+	public Class<? extends String> getReturnType() {

```

Comment:
Why did you change this?

Reply:
I kind of worked from the ground up, so this was unintentional.